File: Backend/src/main/java/letterally/entities/Essay.java
package letterally.entities;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import jakarta.persistence.*;
import lombok.*;
import java.time.LocalDateTime;
import java.util.List;

@Entity
@Table(name = "essays")
@Getter
@Setter
@NoArgsConstructor
@ToString
@JsonIgnoreProperties({"hibernateLazyInitializer","handler"})
public class Essay {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Setter(AccessLevel.NONE)
    private Long id;

    @Column(nullable = false)
    private String title;

    @Column(nullable = false, columnDefinition = "TEXT")
    private String content;

    @Column(name = "created_on", nullable = false)
    private LocalDateTime createdOn;

    @Column(name = "last_updated")
    private LocalDateTime lastUpdated;

    @Column(name = "essay_image")
    private String image;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "topic_id", nullable = false)
    private Topic topic;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @OneToMany(mappedBy = "essay", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Feedback> feedback;

    public Essay(String title, String content, String image, Topic topic, User user) {
        this.title = title.trim();
        this.content = content.trim();
        this.image = image.trim();
        this.topic = topic;
        this.user = user;
    }

    @PrePersist
    public void onCreate() {
        this.createdOn = LocalDateTime.now();
    }

    @PreUpdate
    public void onUpdate() {
        this.lastUpdated = LocalDateTime.now();
    }
}

File: Frontend/src/interfaces/Essay.ts
import type { Feedback } from "./Feedback";
import type { Topic } from "./Topic";
import type { User } from "./User";

export interface Essay {
  id?: number;
  title: string;
  content: string;
  imageFile?: string;
  topic?: Topic;
  user?: User;
  createdOn?: Date;
  votes?: Feedback[];
}

export interface CreateEssayPayload {
  title: string;
  content: string;
  imageFile: File;
  topicId: number;
}

export type UpdateEssayPayload = {
  title: string;
  content: string;
  topicId: number;
  imageFile?: File;
};

export type EssayResponse = {
  id?: number;
  title: string;
  content: string;
  image: string;
  topic?: Topic;
  user?: User;
  createdOn?: Date;
  votes?: Feedback[];
};
